src/common/src/hash/key.rs

```diff
@@ -291,7 +297,18 @@ impl BuildHasher for PrecomputedBuildHasher {
 /// be delegated to other encoding algorithms, we can use macros of
 /// `impl_memcmp_encoding_hash_key_serde!` and `impl_value_encoding_hash_key_serde!` here.
 pub trait HashKeySer<'a>: ScalarRef<'a> {
+    /// Serialize the scalar into the given buffer.
     fn serialize_into(self, buf: impl BufMut);
+
+    /// Returns `Some` if the serialized size is known for this scalar type.
+    fn exact_size() -> Option<usize> {
+        None
+    }
+
+    /// Returns the estimated serialized size for this scalar.
+    fn estimated_size(self) -> usize {
+        Self::exact_size().expect("not exact size")
```

Comment:
```suggestion
        Self::exact_size().unwrap_or(0)
```

Should we just return 0 here? It's just an estimate anyw.
We should rely on `exact_size` to know if there's an exact size or not.

Reply:
Actually, this should be unreachable as we guarantee that either `exact_size` is `Some`, or we manually implement `estimated_size`. But indeed we can write `unwrap_or` here to make it safer.